src/Umbraco.Web/Editors/LanguageController.cs

```diff
@@ -35,59 +42,119 @@ public IEnumerable<Culture> GetAllCultures()
         /// </summary>
         [HttpDelete]
         [HttpPost]
-        public void DeleteLanguage(int id)
+        public IHttpActionResult DeleteLanguage(int id)
         {
             var language = Services.LocalizationService.GetLanguageById(id);
-            if (language == null)
-            {
-                throw new EntityNotFoundException(id, $"Could not find language by id: '{id}'.");
-            }
+            if (language == null) return NotFound();
 
-            if (language.IsDefaultVariantLanguage)
+            var totalLangs = Services.LocalizationService.GetAllLanguages().Count();
+
+            if (language.IsDefaultVariantLanguage || totalLangs == 1)
             {
-                var message = $"Language with id '{id}' is currently set to 'default' and can not be deleted.";
-                throw new HttpResponseException(Request.CreateErrorResponse(HttpStatusCode.BadRequest, message));
+                var message = $"Language '{language.IsoCode}' is currently set to 'default' or it is the only installed language and can not be deleted.";
+                throw new HttpResponseException(Request.CreateNotificationValidationErrorResponse(message));
             }
 
             Services.LocalizationService.Delete(language);
+
+            return Ok();
         }
 
         /// <summary>
-        /// Saves a bulk set of languages with default/mandatory settings and returns the full set of languages configured.
+        /// Creates or saves a language
         /// </summary>
         [HttpPost]
-        public IEnumerable<Language> SaveLanguages(IEnumerable<Language> languages)
+        public Language SaveLanguage(Language language)
         {
-            foreach (var l in languages)
+            if (!ModelState.IsValid)
+                throw new HttpResponseException(Request.CreateValidationErrorResponse(ModelState));
+
+            var found = Services.LocalizationService.GetLanguageByIsoCode(language.IsoCode);
+
+            if (found != null && language.Id != found.Id)
+            {
+                //someone is trying to create a language that alraedy exist
+                ModelState.AddModelError("IsoCode", "The language " + language.IsoCode + " already exists");
+                throw new HttpResponseException(Request.CreateValidationErrorResponse(ModelState));
+            }
+
+            if (found == null)
             {
-                var language = Services.LocalizationService.GetLanguageByIsoCode(l.IsoCode);
-                if (language == null)
+                CultureInfo culture;
+                try
+                {
+                    culture = CultureInfo.GetCultureInfo(language.IsoCode);
+                }
+                catch (CultureNotFoundException)
                 {
-                    language = new Core.Models.Language(l.IsoCode);
+                    ModelState.AddModelError("IsoCode", "No Culture found with name " + language.IsoCode);
```

Comment:
Nice touch - I assume for the cases where for Umbraco Cloud customers have asked to try & setup some more weird & wonderful languages for them

Reply:
I did this to just be safe in case an invalid culture is sent up, since only valid cultures will be sent by Angular anyways since it requests them in the first place.